### PR TITLE
Show actual versions of wrappers and graviton 

### PR DESCRIPTION
**Pull request by PanteraLX**
Aims of this pull request:

* At build time the versions of relevant packages are read with `composer show` and written to a `versions.json` file inside the cache.
* The `X-Version` link contains a summary of the contents of the version file formatted similar to a User-Agent string (ie. `0.9.0; graviton/graviton: 0.25.0; evoja/service-bundle: 0.x.y`)
* These versions are also reported in a new service (which is also exposed in swagger and has a schema)
* The version property of our swagger contains the same data as the X-Version string (unless swagger only supports semver and not strings, then the current wrappers version is to be displayed).

**todo**

* [x] write unit tests for for CoreUtils
* [x] write acceptance tests for the command
* [x] write acceptance tests for the controller
* [x] make sure ``composer check`` does not find heaps of errors
* [x] ensure this runs through on travis
* [x] look into scrutinizer and fix issues you are introducing
* [x] make sure that the above description describes what is actually being implemented
* [x] update docs (seeing this is an important feature for debugging it should prominently be described in the docs at https://gravity-platform-docs.nova.scapp.io/api/)

This is  [EVO-403](https://issue.swisscom.ch/browse/EVO-403)...

### CODE REVIEW

**Comment by hairmare** [comment]
Please run phpunit and composer check on a regular basis... They both show issues that probably should not make their way into a PR.

File: src/Graviton/SwaggerBundle/Service/Swagger.php [hunk 1]
```diff
@@ -174,11 +183,13 @@ private function getBasicStructure()
...
         $ret['swagger'] = '2.0';
+        $date = date('Y-m-d');
         $ret['info'] = array(
             // @todo this should be a real version - but should it be the version of graviton or which one?
-            'version' => '0.1',
+            'version' => $this->coreUtils->getVersionById('graviton')['version'],
             'title' => 'Graviton REST Services',
-            'description' => 'Testable API Documentation of this Graviton instance.'
+            'description' => 'Testable API Documentation of this Graviton instance.',
+            'lastUpdate' => $date
```
Comment (hairmare) [inline src/Graviton/SwaggerBundle/Service/Swagger.php]: We need to double-check that this doesn't lead to our swagger being inavlid...

File: composer.json [hunk 2]
```diff
@@ -73,6 +73,7 @@
             "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::installRequirementsFile",
             "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::removeSymfonyStandardFiles",
             "Graviton\\SwaggerBundle\\Composer\\ScriptHandler::generateSwaggerJson"
+
```
Comment (hairmare) [inline composer.json]: remove this  line from changes

File: src/Graviton/CoreBundle/Compiler/VersionCompilerPass.php [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+    /**
+     * @param string $rootDir path to root dir
+     *
+     * @return array version numbers of packages
+     */
+    private function getPackageVersions($rootDir)
+    {
+        $versions = array();
+        array_push($versions, $this->getContextVersion($rootDir));
```
Comment (hairmare) [inline src/Graviton/CoreBundle/Compiler/VersionCompilerPass.php]: It would be much more concise to just write this instead:

```php
$versions = [
  $this->getContextVersion($rootDir),
];
```

File: src/Graviton/CoreBundle/Compiler/VersionCompilerPass.php [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    private function getContextVersion($rootDir)
+    {
+        if (strpos($rootDir, 'vendor')) {
+            //@todo shell exec needs to be refactored
+            $result = shell_exec('cd ' . escapeshellarg($rootDir) . '/../../../../  && composer show -s --no-ansi');
+        } else {
+            $result = shell_exec('composer show -s --no-ansi');
+        }
+        $lines = explode(PHP_EOL, $result);
+        $wrapper = array();
```
Comment (hairmare) [inline src/Graviton/CoreBundle/Compiler/VersionCompilerPass.php]: It's currently possible that this methods returns an empty array. Is that intended?

File: src/Graviton/CoreBundle/Compiler/VersionCompilerPass.php [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+            $packageNames = shell_exec($command);
+        } else {
+            $packageNames = shell_exec('composer show -i');
+        }
+        $packages = explode(PHP_EOL, $packageNames);
+        //last index is always empty
+        array_pop($packages);
+        foreach ($packages as $package) {
+            preg_match_all('/([^\s]+)/', $package, $match);
+            if ((strpos($match[0][0], 'grv') === 0 | $match[0][0] === 'graviton') && !empty($match[0][1])) {
```
Comment (hairmare) [inline src/Graviton/CoreBundle/Compiler/VersionCompilerPass.php]: And what if I want to add ``symfony/symfony`` to the mix?

I'm not happy with this preg_matcher voodoo at all. We should make this nice and configurable somewhere.

I'd rather have to add each and every package that needs listing to a config file than having to ever add something here.

We need to refactor this so it uses an ``app/config/version_service.yml`` file or something along those lines?

File: src/Graviton/CoreBundle/Controller/VersionController.php [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    /**
+     * Returns all version numbers
+     *
+     * @param Request $request Current http request
+     *
+     * @return \Symfony\Component\HttpFoundation\Response $response Response with result or error
+     */
+    public function allAction(Request $request)
```
Comment (hairmare) [inline src/Graviton/CoreBundle/Controller/VersionController.php]: s/allAction/getAction/

File: src/Graviton/CoreBundle/GravitonCoreBundle.php [hunk 7]
```diff
@@ -58,4 +60,18 @@ public function getBundles()
             new GravitonFileBundle(),
         );
     }
+
+    /**
+     * load compiler pass rest route loader
```
Comment (hairmare) [inline src/Graviton/CoreBundle/GravitonCoreBundle.php]: update docblocks if you really need to copy-paste them

File: src/Graviton/CoreBundle/Service/CoreUtils.php [hunk 8]
```diff
@@ -15,29 +15,71 @@
...
+     */
+    public function getVersion()
     {
-        //@todo if we're in a wrapper context, use the version of the wrapper, not graviton
-        $composerFile = !empty($composerFile) ? $composerFile : __DIR__ . '/../../../../composer.json';
+        $ver = array();
+        foreach ($this->versions as $version) {
+            $ver[$version['id']]= $version['version'];
+        }
+            return $ver;
```
Comment (hairmare) [inline src/Graviton/CoreBundle/Service/CoreUtils.php]: weird indenting is weird

File: src/Graviton/CoreBundle/Service/CoreUtils.php [hunk 9]
```diff
@@ -15,29 +15,71 @@
...
+    }
 
-        if (file_exists($composerFile)) {
-            $composer = json_decode(file_get_contents($composerFile), true);
+    /**
+     * @param string $id add package name
+     * @return array single entry
+     */
+    public function getVersionById($id)
```
Comment (hairmare) [inline src/Graviton/CoreBundle/Service/CoreUtils.php]: I think you can remove this now... At least the possible "This id could not be resolved" thing needs to go.

File: src/Graviton/CoreBundle/Service/CoreUtils.php [hunk 10]
```diff
@@ -15,29 +15,71 @@
...
-                    json_last_error()
-                );
+    }
 
-                throw new \RuntimeException($message);
+    /**
+     *
+     * @return array wrapper version
+     */
+    public function getWrapperVersion()
```
Comment (hairmare) [inline src/Graviton/CoreBundle/Service/CoreUtils.php]: Same here, proabably obsolete.

File: src/Graviton/CoreBundle/Tests/Controller/AppControllerTest.php [hunk 11]
```diff
@@ -46,7 +46,6 @@ public function setUp()
             'doctrine_mongodb'
         );
     }
-
```
Comment (hairmare) [inline src/Graviton/CoreBundle/Tests/Controller/AppControllerTest.php]: please remove whitespace changes in file

File: src/Graviton/CoreBundle/Tests/Services/CoreUtilsTest.php [hunk 12]
```diff
@@ -14,47 +14,48 @@
  */
 class CoreUtilsTest extends \PHPUnit_Framework_TestCase
 {
+    private $versions = array(
+                array("id" => "self", "version" => "0.25.1"),
```
Comment (hairmare) [inline src/Graviton/CoreBundle/Tests/Services/CoreUtilsTest.php]: weird indenting, please use 4 spaces per level and not more.

File: src/Graviton/RestBundle/Tests/Listener/XVersionResponseListenerTest.php [hunk 13]
```diff
@@ -24,7 +24,7 @@ class XVersionResponseListenerTest extends \PHPUnit_Framework_TestCase
     public function testOnKernelResponse()
     {
         $response = new Response();
-        $version = '0.1.0-alpha';
+        $version = '0.25.1';
```
Comment (hairmare) [inline src/Graviton/RestBundle/Tests/Listener/XVersionResponseListenerTest.php]: Why are you changing this test? I  looks a bit smelly that this is the same version that graviton is tagged as in you branch... Please revert the above and maybe use ``disableOriginalConstructor`` instead of ``setConstructorArgs`` when creating a mock below. Doesn't CoreUtils take an array nowadays?

File: src/Graviton/SwaggerBundle/Command/SwaggerGenerateCommand.php [hunk 14]
```diff
@@ -131,7 +131,7 @@ protected function execute(InputInterface $input, OutputInterface $output)
 
         $this->filesystem->dumpFile(
             $this->rootDir.'/../web/swagger.json',
-            json_encode($this->apidoc->getSwaggerSpec())
+            json_encode($this->apidoc->getSwaggerSpec(), JSON_PRETTY_PRINT)
```
Comment (hairmare) [inline src/Graviton/SwaggerBundle/Command/SwaggerGenerateCommand.php]: please revert this change, it should be a pr in it's own right if we really need it, but I don't fell we do since it make an already rather large file even bigger for no apparent reason...

You can always do ``curl https://graviton.nova.scapp.io/swagger.json | python -mjson.tool`` on most platforms if you need it pretty printed.

File: src/Graviton/SwaggerBundle/Service/Swagger.php [hunk 15]
```diff
@@ -174,11 +183,12 @@ private function getBasicStructure()
     {
         $ret = array();
         $ret['swagger'] = '2.0';
+        $date = date('Y-m-d');
```
Comment (hairmare) [inline src/Graviton/SwaggerBundle/Service/Swagger.php]: More like time of last deploy ;)

Please remove the lastUpdate stuff completely for now.

We can do it in another issue if we still want it and it was optional in the original issue for reasons.

**Comment by hairmare** [comment]
I didn't do any review on the schema stuff yet. I'll look into that as soon as we have some schema stuff that's going in the right direction.

File: src/Graviton/CoreBundle/Compiler/VersionCompilerPass.php (hunk 4 above)
Comment (janfriedli) [inline src/Graviton/CoreBundle/Compiler/VersionCompilerPass.php]: The return value is array so if nothing is found it should maybe throw an exception or at least return an empty array?

**Comment by janfriedli** [comment]
Issue [EVO-403](https://issue.swisscom.ch/browse/EVO-403)

File: src/Graviton/CoreBundle/Service/CoreUtils.php (hunk 9 above)
Comment (hairmare) [inline src/Graviton/CoreBundle/Service/CoreUtils.php]: From what I could tell by looking at it this really is only getting called by its tests...

@janfriedli Could you look into this and delete it or let me know where it's being called.

File: src/Graviton/CoreBundle/Service/CoreUtils.php (hunk 10 above)
Comment (janfriedli) [inline src/Graviton/CoreBundle/Service/CoreUtils.php]: This one is called (here)[https://github.com/libgraviton/graviton/blob/5b95bde6d257a146fc833cf030f61d7a1a2a13a9/src/Graviton/SwaggerBundle/Service/Swagger.php#L187]

File: src/Graviton/CoreBundle/Service/CoreUtils.php (hunk 9 above)
Comment (janfriedli) [inline src/Graviton/CoreBundle/Service/CoreUtils.php]: That's true I'll remove it.

**Comment by janfriedli** [comment]
PR which belong to this one
* [wrapper](https://git.swisscom.ch/projects/GRV/repos/graviton-api-szkb-bap/commits?until=refs%2Fheads%2Ffeature%2Fswagger-version-numbers)
* [Doc] (https://github.com/gravity-platform/doc/pull/8)

File: src/Graviton/CoreBundle/Controller/VersionController.php [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+/**
+ * @author   List of contributors <https://github.com/libgraviton/graviton/graphs/contributors>
+ * @license  http://opensource.org/licenses/gpl-license.php GNU Public License
+ * @link     http://swisscom.ch
+ */
+class VersionController extends RestController
+{
+
+    /**
+     * @var coreUtils
```
Comment (hairmare) [inline src/Graviton/CoreBundle/Controller/VersionController.php]: Should be "CoreUtils".

File: src/Graviton/CoreBundle/Compiler/VersionCompilerPass.php [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+ * @license  http://opensource.org/licenses/gpl-license.php GNU Public License
+ * @link     http://swisscom.ch
+ */
+class VersionCompilerPass implements CompilerPassInterface
+{
+
+    /**
+     * @var rootDir
+     */
+    private $rootDir;
```
Comment (hairmare) [inline src/Graviton/CoreBundle/Compiler/VersionCompilerPass.php]: Should be `@var string`.

File: src/Graviton/CoreBundle/Compiler/VersionCompilerPass.php [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+
+    /**
+     * @var config
```
Comment (hairmare) [inline src/Graviton/CoreBundle/Compiler/VersionCompilerPass.php]: Should be `@var array` since you put the results of `getContextVersion` in it.

File: src/Graviton/CoreBundle/Tests/Services/CoreUtilsTest.php [hunk 19]
```diff
@@ -14,47 +14,38 @@
...
-     */
-    public function testGetVersion($version, $filePath = '')
-    {
-        $utils = new CoreUtils();
-        $this->assertEquals($version, $utils->getVersion($filePath));
-    }
+    private $versions = array(
+                array("id" => "self", "version" => "0.25.1"),
+                array("id" => "financing", "version" => "0.1"),
+            );
```
Comment (hairmare) [inline src/Graviton/CoreBundle/Tests/Services/CoreUtilsTest.php]: Needs fixed formatting and a docblock...
